pkgs/os-specific/linux/dcgm/default.nix

```diff
@@ -0,0 +1,108 @@
+{ lib
+, callPackage
+, gcc11Stdenv
+, fetchFromGitHub
+, addOpenGLRunpath
+, catch2
+, cmake
+, cudaPackages_10_2
+, cudaPackages_11_8
+, cudaPackages_12
+, fmt_9
+, git
+, jsoncpp
+, libevent
+, plog
+, python3
+, tclap_1_4
+, yaml-cpp
+}:
+let
+  # Flags copied from DCGM's libevent build script
+  libevent-nossl = libevent.override { sslSupport = false; };
+  libevent-nossl-static = libevent-nossl.overrideAttrs (super: {
+    CFLAGS = "-Wno-cast-function-type -Wno-implicit-fallthrough -fPIC";
+    CXXFLAGS = "-Wno-cast-function-type -Wno-implicit-fallthrough -fPIC";
+    configureFlags = super.configureFlags ++ [ "--disable-shared" "--with-pic" ];
+  });
+
+  jsoncpp-static = jsoncpp.override { enableStatic = true; };
+
+  cudaToolkits = [
+    {
+      version = "10";
+      toolkit = cudaPackages_10_2.cudatoolkit;
+    }
+    {
+      version = "11";
+      toolkit = cudaPackages_11_8.cudatoolkit;
+    }
+    {
+      version = "12";
+      toolkit = cudaPackages_12.cudatoolkit;
+    }
+  ];
+
+  # Builds CMake code to add CUDA paths for include and lib.
+  mkAppendCudaPaths = { version, toolkit }: ''
+    list(APPEND Cuda${version}_INCLUDE_PATHS "${toolkit}/include")
+    list(APPEND Cuda${version}_LIB_PATHS "${toolkit}/lib" "${toolkit.lib}/lib")
+  '';
+in gcc11Stdenv.mkDerivation rec {
+  pname = "dcgm";
+  version = "3.1.8";
+
+  src = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "DCGM";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-OXqXkP2ZUNPzafGIgJ0MKa39xB84keVFFYl+JsHgnks=";
+  };
+
+  # Add our paths to the CUDA paths so FindCuda.cmake can find them.
+  EXTRA_CUDA_PATHS = lib.strings.concatMapStringsSep "\n" mkAppendCudaPaths cudaToolkits;
+  prePatch = ''
+    echo "$EXTRA_CUDA_PATHS"$'\n'"$(cat cmake/FindCuda.cmake)" > cmake/FindCuda.cmake
+  '';
+
+  hardeningDisable = [ "all" ];
+
+  nativeBuildInputs = [
+    addOpenGLRunpath
+    cmake
+    git
+    python3
+
+    jsoncpp-static
+    jsoncpp-static.dev
+    libevent-nossl-static
+    libevent-nossl-static.dev
+    plog.dev # header-only
+    tclap_1_4 # header-only
+  ];
+
+  buildInputs = [
+    catch2
+    fmt_9
+    yaml-cpp
+  ];
+
+  # libcuda.so must be found at runtime because it is supplied by the NVIDIA
+  # driver.
+  postFixup = ''
+    find "$out/bin" "$out/lib" -type f -executable -print0 | while IFS= read -r -d "" f; do
+      if isELF "$f" && [[ $(patchelf --print-needed "$f" || true) == *libcuda.so* ]]; then
+        addOpenGLRunpath "$f"
+      fi
+    done
+  '';
```

Comment:
should be able to replace all of this with autoAddOpenGLRunpathHook. See eg https://github.com/NixOS/nixpkgs/blob/3297ece10cb7cdcc5c166f2fdaeb8fc5f84d3ec6/pkgs/applications/science/misc/colmap/default.nix#L39

Reply:
I realize it doesn't *currently* matter because of the way this hook is written, but that would require that I use a hook from a *particular* version cudaPackage, when the DCGM build system explicitly uses 3 of them. I did not want to be biased.

Reply:
ah yes this is an unfortunate consequence of how `autoAddOpenGLRunpathHook` is packaged. for better or worse `autoAddOpenGLRunpathHook` is the same in each `cudaPackages` attr set. I would recommend just picking the latest one and pulling from there

Reply:
depending on more than one `cudaPackages` in a single derivation is quite out of the ordinary :P

Reply:
Fair enough. I can switch to the hook and just add a comment that it doesn't *actually* pick a version for you. Yes, I was very surprised to find that DCGM needs all three and I did try to change their build system to avoid that. However, as noted in my comments to @Kiskae's review, it is *not* a trivial change to their build system to achieve that.

Reply:
The good news is that DCGM's *runtime* dependency is actually pretty small.

Reply:
autoAddOpenGLRunpathHook has a bug when it encounters statically linked things, apparently. I don't want to rebuild the world so I left this in here with a comment. I did switch it in `dcgm-exporter` though.

Reply:
Hmm that's unfortunate... what was the error exactly? could you please create an issue for it? we should really get on top of that

Reply:
I will open a separate PR to fix the hook and remove this hack.

Reply:
even better!